### ISSUE TITLE
Fixed: how to initialize with accuracy retained?

### ISSUE THREAD

**Issue by axefrog** [eteran/cpp-utilities#2]
Hi, I just ran a quick test like this:

```c++
typedef numeric::Fixed<16, 16> fixed;
fixed f = 5.3567;
std::cout << "fixed: " << f << std::endl;
```

The output was:

```
fixed: 5.35669
```

Could you help me to understand the discrepancy I'm seeing?

**Comment by eteran** [comment]
Sure, let me look into it and see what's going on. I suspect that the `5.3567` can't be stored precisely in a `float` making the issue a matter of how to get the accurate value in, in the first place :-)

**Comment by axefrog** [comment]
My first instinct was actually to look for a constructor like this:
```c++
fixed f(5, 3567);
```
Alas, it was not there!

**Comment by eteran** [comment]
OK, so this is an issue with how the values are stored and perhaps there simply not being enough fractional bits to accurately represent that particular value. Here's what's happening:

* `1 << 16 = 65,536` (this is the value of "`1`" in a 16.16 FP type
* `5.3567 * 65,536 = 351,056.6912` (we scale your input value based on that)
* `351,056.6912` truncated = `351,056` (and store it as a `uint32_t`)
* `351,056` = `0x55B50`

Here's what we get:
* integer value = `0x05` (good)
* fractional value = `0x5B50 / 65,536` = `0.356689...` which is exactly what you see! (but not good)
	
So, the first thing that comes up is that a `Fixed<32,32>` **will** accurately store that value. I like the idea of your constructor though to pass in the parts piece wise, I think I'll add that. I will look into if that fraction is even possible to represent in a 16.16 FP (I fear not, but doesn't hurt to try :-))

**Comment by eteran** [comment]
If you do simply go with `Fixed<32,32>`, please note that division requires a `128-bit` type to be accurate and only is tested to work well on gcc/clang

**Comment by eteran** [comment]
Also, as a note, unfortunately now that I've looked at it, a `Fixed(int i, int f)` constructor doesn't seem doable in any practical sense. The reason why is that I have no way of knowing how much to scale the fractional part!

If the user writes `fixed f(5, 3567);` Do they mean `5.3567` or `5.03567` or `5.003567`, etc...

Essentially, I need to scale that integer by `65,536`, but I only want the top portion (if they write `3567890`,  `3567`  is more important to keep that `890` since that is more significant to the value. But I have no way of knowing that the best value to scale by (in this case is `1000`). The bigger the number, the more in needs scaling. As I type this, I am imagining some `log10` based method to figure it out. But it'll be messy.

In the end, a better option would simply let the user put in a string, which is parsed and stored appropriately :-) (which I may add...)

**Comment by axefrog** [comment]
Ah, thanks for the breakdown! What about an extra constructor parameter to indicate leading zeros?
```c++
fixed f(5, 3567); // 5.3567
fixed f(5, 3567, 0); // 5.3567
fixed f(5, 3567, 2); // 5.003567
```

**Comment by axefrog** [comment]
By the way... don't necessarily do this on my account; I'm still prototyping and haven't even yet figured out if this is the right approach.

**Comment by eteran** [comment]
Yea, that would be doable. Though in the end, when looking at it. I don't believe there is any 16.16 bit pattern that will retain accuracy. If the integer portion will be relatively small, you could use an 8.24, which will give you more (but never 100%) accuracy. Or like I said, also a 32.32 will do the trick.

Another alternative is to use a **Decimal** based system. (this is a fixed point binary lib). So for example, when you want to represent: `5.3567` you simply store "53567" in your `uint32_t` directly. Divide by `10000` to get the integer part, Mod `10000` to get the fractional (accurate to 4 digits) part.

Of course division will always make it possible to have inaccuracies (like 1/3 = 0.3333....), but it's good if you know how many fractional digits you care about (Decimal is of course critical for $ calcs because of this)

**Comment by eteran** [comment]
Heh, so I just did a test, you are literally 1 bit shy of having enough to represent that value a `Fixed<15.17>` prints the expected number. Bear in mind, that this is enough bits for THIS value. There will always be some values which simply must be approximated. So, closing this issue, but feel free to continue to discussion here if it helps.

**Comment by axefrog** [comment]
That's cool, thanks for the discussion. I'm thinking about this stuff from the start because my game server is designed to be distributed, but there will be a lot of procedurally-generated data, so not considering this stuff up front could result in some pretty nasty bugs down the track.
